Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,647 @@
+| -------------------- | --------------------- | ------------------------------------------------------------- |
+| `*` / `*?`           | 0 or more             | `Array` of the sub-pattern capture type                       |
+| `+` / `+?`           | 1 or more             | `Array` of the sub-pattern capture type                       |
+| `?` / `??`           | 0 or 1                | `Optional` of the sub-pattern capture type                    |
+| `{n}` / `{n}?`       | Exactly _n_           | `Array` of the sub-pattern capture type                       |
+| `{n, m}` / `{n, m}?` | Between _n_ and _m_   | `Array` of the sub-pattern capture type                       |
+| `{n,}` / `{n,}?`     | _n_ or more           | `Array` of the sub-pattern capture type                       |
+
+```swift
+/([0-9a-fA-F]+)+/
+// => `Regex<[Substring]>`
+
+// Result builder DSL equivalent:
+//     OneOrMore {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     }
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+)*/
+// => `Regex<[Substring]>`
+
+// Result builder DSL equivalent:
+//     Repeat {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     }
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+)?/
+// => `Regex<Substring?>`
+
+// Result builder DSL equivalent:
+//     Optionally {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     }
+//     // `.Captures == Substring?`
+
+/([0-9a-fA-F]+){3}/
+// => `Regex<[Substring]>
+
+// Result builder DSL equivalent:
+//     Repeat(3) {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     )
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+){3,5}/
+// => `Regex<[Substring]>`
+
+// Result builder DSL equivalent:
+//     Repeat(3...5) {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     )
+//     // `.Captures == [Substring]`
+
+/([0-9a-fA-F]+){3,}/
+// => `Regex<[Substring]>`
+
+// Result builder DSL equivalent:
+//     Repeat(3...) {
+//         OneOrMore(CharacterClass.hexDigit).capture()
+//     )
+//     // `.Captures == [Substring]`
+```
+
+Note that capturing collections of repeated captures like this is a departure
+from most regular expression implementations, which only provide access to the
+_last_ match of a repeated capture group. For example, Python only captures the
+last group in this dash-separated string:
+
+```python
+rep = re.compile('(?:([0-9a-f]+)-?)+')
+match = rep.match("1234-5678-9abc-def0")
+print(match.group(1))
+# Prints "def0"
+```
+
+By contrast, the proposed Swift version captures all four sub-matches:
+
+```swift
+let pattern = /(?:([0-9a-f]+)-?)+/
+if let match = “1234-5678-9abc-def0".firstMatch(of: pattern) {
+    print(match.captures)
+}
+// Prints [“1234”, “5678", “9abc”, “def0"]
+```
+
+Despite the deviation from prior art, we believe that the proposed capture
+behavior leads to better consistency with the meaning of these quantifiers.
```

Comment:
But we can also provide a compatibility projection. BTW, do we know why they only provide the last one? Do we want that kind of memory savings or does that simplify our execution strategy?

Reply:
I added a paragraph about adding a variant of quantifiers for smaller memory footprint as a future direction. d8ccc63